### ISSUE TITLE
Overly intense fringes

### ISSUE THREAD

**Issue by 0xBRM** [giddie/poppler-cairo-backend#7]
[Screenshot comparison](http://screenshotcomparison.com/comparison/166411)
Look at the t's, the i's and the p's. I know it is required but it seems a bit excessive.

Zoomed in is [nowhere near as bad](http://screenshotcomparison.com/comparison/166412).

Edit: for some reason it seems to alter Okular's thumbnail generation as well, and not simply the text based ones.
Edit2: I am using poppler 0.42.0.

**Comment by giddie** [comment]
Yeah, I'm not happy with the fringes either. I imagine there's some filtering that could be applied to this, but I'm afraid I know pretty much nothing about how Cairo does subpixel rendering. Patches or research would be greatly appreciated. A good starting point might be to look for any patches people have made to apply subpixel rendering in Evince, as we should be able to apply the same code in this patchset.

It doesn't surprise me too much that you're seeing differences in thumbnail generation in Okular: this patchset changes the whole rendering engine that Okular uses to match that of Evince and any other apps that use the Cairo backend.

**Comment by 0xBRM** [comment]
Okay, so the problem is that Cairo, as used by poppler, will default to FT_LCD_FILTER_LEGACY and this is what's causing all the issues. When you force it to use FT_LCD_FILTER_DEFAULT (where the default is "08 24 36 24 08", as per infinality/fontconfig-ultimate's patchset, though you can try using vanilla freetype) the results look much better.

![FT_LCD_FILTER_LEGACY](https://cloud.githubusercontent.com/assets/9098352/13992474/29a03ea6-f114-11e5-9849-dc2096b68b7b.png)

![FT_LCD_FILTER_DEFAULT](https://cloud.githubusercontent.com/assets/9098352/13992492/38ec357c-f114-11e5-98cf-83b8272a61d0.png)

The overly bold text can be remedied by using a more up-to-date pixman, or by including a slightly lighter default LCD filter (LCD_FILTER_LIGHT).

**Comment by giddie** [comment]
So do you know what cairo method call is required to apply the filter? I'm hoping it can be easily added in the Cairo font options in `CairoOutputDev::setCairo()`.

**Comment by 0xBRM** [comment]
It used to be  `cairo_font_options_set_lcd_filter()` but they have since removed it, so I'm using a [patched Cairo](https://github.com/CrisBRM/fontconfig-ultimate/blob/master/cairo/cairo-lcdfilter-make-default-default.patch).

[Screenshot comparison](http://screenshotcomparison.com/comparison/166497)

As you can see it looks noticeably better, without the fringes. Smaller fonts look amazing as well.

**Comment by zhou13** [comment]
You may also interest in my patchset https://github.com/zhou13/poppler-subpixel, which only enables subpixel rendering when the correctness is guaranteed by calling `poppler_page_support_subpixel_rendering`.

**Comment by 0xBRM** [comment]
@zhou13 Ah, I actually used your patchset. 
I'm going to try your patch and report back with results.

Redacted due to misunderstanding.

**Comment by CrisBRM** [comment]
4 comparisons here, using poppler 0.42, patched cairo 1.14.6, fontconfig-ultimate, and the ubuntu font.

#1 - 100% zoom, giddie vs zhou13 (no hinting)
#2 -  100% zoom, giddie vs zhou13 (hinting)
#3 - fit width, giddie vs zhou13 (no hinting)
#4 -  fit width, giddie vs zhou13 (hinting)

http://screenshotcomparison.com/comparison/166917/

They look different and zhou13's don't change the thumbnails for some reason.

**Comment by zhou13** [comment]
What I notice is that zhou13's version  in you screenshot does not has subpixel rendering at all...  Directly applying my patchset won't change the result of okular because it is designed for evince now.  In order to make it work for okular, you need to patch okular also (to explicit enable subpixel rendering).  As for the result, I expect they should be exactly the same on most PDF files unless blending mode are used.

**Comment by 0xBRM** [comment]
@zhou13 Oh, I see, so it doesn't add support for the cairo backend to the Qt4 bindings. My bad. Should have read the patchset.

**Comment by giddie** [comment]
Thanks @zhou13; you've done the Cairo-backend-side work to support subpixel-rendering properly, which is what I was hoping for. It's unfortunate that it requires a patched Cairo. If that weren't the case, I'd certainly try to include your work in this patchset too, as the current patch I use to make the Cairo backend do subpixel-rendering is pretty basic.

Do you know whether Cairo is interested in merging the necessary code upstream?

**Comment by zhou13** [comment]
https://bugs.freedesktop.org/show_bug.cgi?id=10301

It seems that comments in bugzilla are ignored as usual.  Maybe we need to submit it to the developer mail list.
